stdlib/public/core/PrefixWhile.swift

```diff
@@ -99,70 +100,110 @@ extension LazySequenceProtocol {
   }
 }
 
-//===--- Collections ------------------------------------------------------===//
-
-/// A position in the base collection of a `LazyPrefixWhileCollection` or the
-/// end of that collection.
+/// A lazy `${Collection}` wrapper that includes the initial consecutive
+/// elements of an underlying collection that satisfy a predicate.
+///
+/// - Note: The performance of accessing `endIndex`, `last`, any methods that
+///   depend on `endIndex`, or moving an index depends on how many elements
+///   satisfy the predicate at the start of the collection, and may not offer
+///   the usual performance given by the `Collection` protocol. Be aware,
+///   therefore, that general operations on `${Self}` instances may not have
+///   the documented complexity.
 @_fixed_layout // FIXME(sil-serialize-all)
-public enum _LazyPrefixWhileIndexRepresentation<Base : Collection> {
-  case index(Base.Index)
-  case pastEnd
-}
+public struct LazyPrefixWhileCollection<Base: Collection> {
+  public typealias Element = Base.Element
+  
+  @_inlineable // FIXME(sil-serialize-all)
+  @_versioned // FIXME(sil-serialize-all)
+  internal init(_base: Base, predicate: @escaping (Element) -> Bool) {
+    self._base = _base
+    self._predicate = predicate
+  }
 
-/// A position in a `LazyPrefixWhileCollection` or
-/// `LazyPrefixWhileBidirectionalCollection` instance.
-@_fixed_layout // FIXME(sil-serialize-all)
-public struct LazyPrefixWhileIndex<Base : Collection> : Comparable {
-  /// The position corresponding to `self` in the underlying collection.
   @_versioned // FIXME(sil-serialize-all)
-  internal let _value: _LazyPrefixWhileIndexRepresentation<Base>
+  internal var _base: Base
+  @_versioned // FIXME(sil-serialize-all)
+  internal let _predicate: (Element) -> Bool
+}
 
-  /// Creates a new index wrapper for `i`.
+extension LazyPrefixWhileCollection: Sequence {
+  public typealias Iterator = LazyPrefixWhileSequence<Base>.Iterator
+  
   @_inlineable // FIXME(sil-serialize-all)
-  @_versioned // FIXME(sil-serialize-all)
-  internal init(_ i: Base.Index) {
-    self._value = .index(i)
+  public func makeIterator() -> Iterator {
+    return Iterator(_base: _base.makeIterator(), predicate: _predicate)
   }
+}
 
-  /// Creates a new index that can represent the `endIndex` of a
-  /// `LazyPrefixWhileCollection<Base>`. This is not the same as a wrapper
-  /// around `Base.endIndex`.
-  @_inlineable // FIXME(sil-serialize-all)
-  @_versioned // FIXME(sil-serialize-all)
-  internal init(endOf: Base) {
-    self._value = .pastEnd
+extension LazyPrefixWhileCollection {
+  /// A position in the base collection of a `LazyPrefixWhileCollection` or the
+  /// end of that collection.
+  @_fixed_layout // FIXME(sil-serialize-all)
+  @_versioned
+  internal enum _IndexRepresentation {
```

Comment:
We can reuse the `ClosedRange.Index` enum here. It's essentially the same, just needs to have it's requirements relaxed a little bit (see https://github.com/apple/swift/pull/14288/commits/8b54a33969760c6c5b5b2afd81618890b4503581).

Can you add a `FIXME` please? And maybe a radar/jira.

Reply:
OK, I'll merge this and then do that in the other PR, which I need to do another push to.